### ISSUE TITLE
Possible typo in answer to Question 17(b)

### ISSUE THREAD

**Issue by rama100** [rasbt/MachineLearning-QandAI-book#18]
Q: Can we finetune a decoder-only model like GPT for classification?

A: ... "In contrast to BERT, we do not need to use a special [CLS], but the fundamental concept is similar to finetuning an encoder-style model for classification. We add a classification head (a fully connected layer and a softmax activation) and train it on the embedding (the final hidden state) of the **first** output token. (This is analogous to using the [CLS] token embedding.)"

Sebastian: Due to the causal attention mask in the decoder, only the last token's hidden state will be based on all the tokens in the sentence (which is what we want for classification tasks). So, I assume you meant to write "last" (not "first") above? Sorry if I misunderstood.

Wonderful book, BTW. I am really enjoying it!

**Comment by rasbt** [comment]
Thanks for the comment @rama100 . I think that both ways should work since we are providing the whole input in classification tasks, but maybe "last" is more intuitive. I just updated it in the book.
Thanks for the feedback!

**Comment by rama100** [comment]
Thanks!

I must confess I don't see how both would work. We _are_ providing the whole input but **assuming the causal mask in the decoder is being used**,  I would think that :
- the final hidden state corresponding to the _first_ input token is ultimately a function of just the _first_ token and none of the other input tokens, 
- the final hidden state corresponding to the _second_ input token is ultimately a function of just the _first two_ tokens and none of the other input tokens .. 
- ...
- the final hidden state corresponding to the _last_ input token is ultimately a function of _all_ the input tokens?

If the above line of reasoning is correct, only the last token's final hidden state would work as intended. Let me know where I am going wrong. Thanks.

**Comment by rasbt** [comment]
I agree with you for the pretraining scenario where we generate a token at a time starting from the first token. Or if we use the LLM for inference with only a single starting token. But suppose we are to classify or categorize texts where we have the input already given. E.g., I created some arbitrary example below where the model sees "The red car is fast" and starts generating after the <gen> token:

```
Input:     The   red   car    is     fast   <gen>   word1   word2   word3
            ------------------------------------------------------------
The         | .    m     m      m      m      m       m      m      m
red         | .    .     m      m      m      m       m      m      m
car         | .    .     .      m      m      m       m      m      m
is          | .    .     .      .      m      m       m      m      m
fast        | .    .     .      .      .      m       m      m      m
<gen>       | .    .     .      .      .      .       m      m      m
word1       | .    .     .      .      .      .       .      m      m
word2       | .    .     .      .      .      .       .      .      m
word3       | .    .     .      .      .      .       .      .      .
```

So in this case, the first generated token ("word") would have had access to the entire input sequence.

**Comment by rama100** [comment]
Oh, you meant the final hidden state of the first **generated** token, not the final hidden state of the first (input) token!!

This clarifies everything. In fact, on re-reading your original answer, I think you had it right all along:
"We add a classification head (a fully connected layer and a softmax activation) and train it on the embedding (the final hidden state) of the first **output** token."

Thanks for clarifying and taking the time to create the example.

**Comment by rama100** [comment]
> it could be that the first generated output token is slightly preferred as it really only captures the original input context and nothing of the generated text

Agreed.

I think the small change below would help make it clear to the reader.

Instead of this:
"We add a classification head (a fully connected layer and a softmax activation) and train it on the embedding (the final hidden state) of the first **output** token."

This:
"We add a classification head (a fully connected layer and a softmax activation) and train it on the embedding (the final hidden state) of the first **generated output** token."

**Comment by rasbt** [comment]
Ah yes, adding "generated" should help to make it more clear. Thanks for the feedback!
